### PR TITLE
Add separate button "Copy Debug Info"

### PR DESCRIPTION
**Pull request by OttoAllmendinger**
bf9f64c (Otto Allmendinger, 12 days ago)
   Rename to copyDebugInfo()

   Make `error` optional.

   Issue: BG-26842

e94b7e3 (Otto Allmendinger, 12 days ago)
   App reformat

   For some reason IntelliJ insists on that order

   Issue: BG-26842

2ad51ca (Otto Allmendinger, 13 days ago)
   Add "copy error info" button

   Copies error info to clipboard.

   Reorganize code a bit to have `try {}` block smaller.

   Issue: BG-26842

c83a77f (Otto Allmendinger, 13 days ago)
   Add getEmptyState()

   Make it easier to modify initial state.

d4fbc8f (Otto Allmendinger, 13 days ago)
   Add getRecoveryDebugInfo()

45e25dc (Otto Allmendinger, 13 days ago)
   Simplify recoverWithKeyPath

eeaa137 (Otto Allmendinger, 13 days ago)
   logToConsole: use `console.error()`

b39b65c (Otto Allmendinger, 13 days ago)
   Add `getBitgoInstanceFromEnv()`

   Allows more convenient development

   Issue: BG-26842

### CODE REVIEW

**Comment by matthewzipkin-bitgo** [comment]
Testing locally...

Nice debug button!

Screen Shot 2020-12-10 at 3 14 23 PM

Got a few samples:

```
{
  "errorMessage": "",
  "errorStack": "",
  "recoveryDebugInfo": {}
}
{
  "errorMessage": "missing userKey",
  "errorStack": "Error: missing userKey\n    at _callee2$ (http://localhost:3000/bundle.js:541933:19)\n    at tryCatch (http://localhost:3000/bundle.js:155264:40)\n    at Generator.invoke [as _invoke] (http://localhost:3000/bundle.js:155498:22)\n    at Generator.prototype.<computed> [as throw] (http://localhost:3000/bundle.js:155316:21)\n    at step (http://localhost:3000/bundle.js:541822:191)\n    at http://localhost:3000/bundle.js:541822:402",
  "recoveryDebugInfo": {}
}
{
  "errorMessage": "invalid recoveryDestination",
  "errorStack": "Error: invalid recoveryDestination\n    at _callee2$ (http://localhost:3000/bundle.js:541933:19)\n    at tryCatch (http://localhost:3000/bundle.js:155264:40)\n    at Generator.invoke [as _invoke] (http://localhost:3000/bundle.js:155498:22)\n    at Generator.prototype.<computed> [as throw] (http://localhost:3000/bundle.js:155316:21)\n    at step (http://localhost:3000/bundle.js:541822:191)\n    at http://localhost:3000/bundle.js:541822:402",
  "recoveryDebugInfo": {}
}
```

...reviewing code now but LGTM so far 👍

File: src/utils.js [hunk 1]
```diff
@@ -15,18 +21,12 @@ import * as Errors from 'bitgo/dist/src/errors';
...
-    recoveryParams['userKeyPath'] = path;
-
     try {
-      recoveryPrebuild = await baseCoin.recover(recoveryParams);
       // if we already have a recovery result, then it means the current path we try
       // is the valid user path, and we can return and exit the iteration loop
-      if (recoveryPrebuild) {
-        return recoveryPrebuild;
-      }
+      return await baseCoin.recover(withCustomKeyPath(recoveryParams, path));
```
Comment (matthewzipkin-bitgo) [inline src/utils.js]: Not sure but you might not need `await` here. This function is already marked `async` and if `baseCoin.recover()` returns a promise then I think you're ok.

File: src/components/non-bitgo.js [hunk 2]
```diff
@@ -105,84 +116,108 @@ class NonBitGoRecoveryForm extends Component {
...
+
+    const fileName = baseCoin.getChain() + '-recovery-' + Date.now().toString() + '.json';
+    const dialogParams = {
+      filters: [
+        {
+          name: 'Custom File Type',
+          extensions: ['json'],
+        },
+      ],
+      defaultPath: '~/' + fileName,
```
Comment (matthewzipkin-bitgo) [inline src/components/non-bitgo.js:183]: Might wanna use `os.homedir()` for this: https://nodejs.dev/learn/the-nodejs-os-module#oshomedir
Not sure if `~/` is easy for users to find on every operating system (looking at you, Windows!)

File: src/App.js [hunk 3]
```diff
@@ -1,5 +1,5 @@
 import React, { Component } from 'react';
-import { BrowserRouter as Router, Switch, Route, Redirect } from 'react-router-dom';
+import { BrowserRouter as Router, Redirect, Route, Switch } from 'react-router-dom';
```
Comment (matthewzipkin-bitgo) [inline src/App.js:2]: Alphabetical? lol that is PICKY! :-P

File: src/App.js [hunk 4]
```diff
@@ -7,9 +7,25 @@ import Login from 'views/login';
...
+  }
+  return new BitGo({ env, accessToken: token });
+}
 
 class App extends Component {
-  state = { isLoggedIn: false, loginBypass: false, bitgo: null };
+  constructor() {
+    super();
+    const bitgo = getBitgoInstanceFromEnv();
+    this.state = bitgo
```
Comment (matthewzipkin-bitgo) [inline src/App.js:25]: What's this for? Doesn't seem relevant to this PR but I do see some calls around the repo that pull `Bitgo` from the state. How did it work before?

Comment (glowang) [reply src/App.js:25]: Looks like `bitgo` instance was assigned during login, stored in the `this.props` object. For example, here is how `bitgo` is assigned for nonBitGo features https://github.com/BitGo/wallet-recovery-wizard/blob/1d29e4812adeff2119f14e7595f46266ccba1da5/src/views/login/index.js#L44  

Although I am not sure how `bypassLogin` does the actual assignment. I don't see any function definition besides 
```
      return <Login finishLogin={this.updateLoginState(true)} bypassLogin={this.updateLoginBypass} {...props} />;
``

**Review by glowang — COMMENTED** [review]
image

Could you run the non bitgo recovery feature locally? I am getting this error.

File: src/utils.js [hunk 5]
```diff
@@ -1,9 +1,29 @@
...
+    }
+
+    return k.neutered().toBase58();
+  });
+}
+
+export async function getRecoveryDebugInfo(baseCoin, recoveryParams) {
+  if (!(baseCoin instanceof AbstractUtxoCoin)) {
+    // TODO support more coins
+    throw new Error('unsupported coin');
```
Comment (glowang) [inline src/utils.js:24]: We should use the coin-config as a point of reference, as we support different coins for testnet and mainnet as well (due to some testnet block explorers being unavailable) 

See https://github.com/BitGo/wallet-recovery-wizard/blob/1d29e4812adeff2119f14e7595f46266ccba1da5/src/constants/coin-config.js#L133

File: src/components/non-bitgo.js [hunk 6]
```diff
@@ -10,26 +10,31 @@ import tooltips from 'constants/tooltips';
...
     backupKey: '',
     bitgoKey: '',
-    rootAddress: '',
+    coin: 'btc',
     walletContractAddress: '',
+    rootAddress: '',
     tokenAddress: '',
     walletPassphrase: '',
     recoveryDestination: '',
-    apiKey: '',
```
Comment (glowang) [inline src/components/non-bitgo.js:28]: This got deleted as a result of refactoring; but we still need this field.

File: src/components/non-bitgo.js [hunk 7]
```diff
@@ -10,26 +10,31 @@ import tooltips from 'constants/tooltips';
...
-    scan: 20,
```
Comment (glowang) [inline src/components/non-bitgo.js:29]: same for `scan`

File: src/App.js (hunk 4 above)
Comment (glowang) [reply src/App.js:25]: Actually, `state` is more appropriate here. We do mutate `bitgo` at some point in non-bitgo

Comment (OttoAllmendinger) [reply src/App.js:25]: this change allows me to set the `REACT_APP_` envars in order to skip the login screen and have a valid bitgo instance

Comment (OttoAllmendinger) [reply src/App.js:25]: since `bitgo` was a `state` property before I didn't change it (just changed the location where `state` is set)

File: src/components/non-bitgo.js (hunk 2 above)
Comment (OttoAllmendinger) [reply src/components/non-bitgo.js:183]: You are correct, but I am just moving around existing code here. We should tackle these sorts of issues in separate PRs.

File: src/utils.js (hunk 5 above)
Comment (OttoAllmendinger) [reply src/utils.js:24]: In this case though we only worry about the coins that `getRecoveryDebugInfo()` method works for.

Since I can't guarantee that the result of `initiateRecovery` works with other coins I narrowed it down to utxo-like coins. It works for all utxo coins though.

File: src/App.js (hunk 3 above)
Comment (OttoAllmendinger) [reply src/App.js:2]: somehow `prettier` works slightly differently when used from IntelliJ :man_shrugging:

File: src/utils.js (hunk 1 above)
Comment (OttoAllmendinger) [reply src/utils.js]: `await` is necessary here in order for the `try` block in this function to catch an exception thrown during promise resolution

Comment (matthewzipkin-bitgo) [reply src/utils.js]: 👍

File: src/components/non-bitgo.js (hunk 2 above)
Comment (matthewzipkin-bitgo) [reply src/components/non-bitgo.js:185]: 👍

File: src/App.js (hunk 4 above)
Comment (matthewzipkin-bitgo) [reply src/App.js:25]: ah gotcha, cool 👍

**Review by matthewzipkin-bitgo — APPROVED** [review]

**Comment by glowang** [comment]
I tested a nonBitgo recovery with mainnet BTC, and I got this error when I tried to copy debug info:
```
index.js:2178 error gathering recovery debug info Error: unsupported coin
    at _callee$ (utils.js:21)
    at tryCatch (runtime.js:62)
    at Generator.invoke [as _invoke] (runtime.js:296)
    at Generator.prototype.<computed> [as next] (runtime.js:114)
    at step (index.js:2)
    at index.js:2
    at new Promise (<anonymous>)
    at index.js:2
    at getRecoveryDebugInfo (utils.js:21)
    at NonBitGoRecoveryForm._callee$ (non-bitgo.js:73)
    at tryCatch (runtime.js:62)
    at Generator.invoke [as _invoke] (runtime.js:296)
    at Generator.prototype.<computed> [as next] (runtime.js:114)
    at step (migrated-legacy.js:374)
    at migrated-legacy.js:374
```
I suspect there is something funky with the way `instance of` works

**Comment by glowang** [comment]
There is a similar error from the sanitize keys function wrt using `instanceof`. But the derivation of pub keys is successful once I commented off 
```
     if (!(k instanceof utxolib.HDNode)) {
       throw new Error(`unexpected key`);
     }
```
